### DISCUSSION TITLE
Product Button & Interactivity API with cache plugins

### DISCUSSION THREAD

**Discussion by gigitux** [woocommerce/woocommerce#44961] [WooCommerce, Blocks, and Site Editing]
During the review of the PR that introduce the Interactivity API (#10006) to the Product Button @Aljullu found an issue with the cache plugins:

> The only big issue I found is that if I enable a caching plugin, then the session from one user might leak into another. In this screenshot you can see I have only one product in the cart, but the Add to Cart buttons show I have more:
![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/7860d6dc-fff7-487b-b83a-b0ed5689d608)
> [@Aljullu 's comment](https://github.com/woocommerce/woocommerce-blocks/pull/10006#pullrequestreview-1518469781) 

The Interactivity API utilizes the `data-*` attribute to store user data, but it's important to note that cache plugins, such as WP-Optimize, cache the entire HTML code, including the `data-*` attributes. Consequently, there is a risk of user data leakage across different sessions.

To mitigate this risk, WooCommerce has implemented a function that disables cache plugins for Cart, Checkout, and Account pages ([source code](https://github.com/woocommerce/woocommerce/blob/9c80e4d53fd7d24fd962468dd34219647f1a54ff/plugins/woocommerce/includes/class-wc-cache-helper.php/#L135-L148)).

However, with the introduction of the Site Editor, blocks can be placed anywhere on a page, which makes it challenging to rely on page-specific approaches. Additionally, the same block, like the Mini Cart block, can be present on multiple pages. These factors make it more likely for the issue of user data leakage to occur, especially when using the Interactivity API.

Questions to consider:
- Should we disable the cache per-block basis (not sure that it is possible)?
- Should we take care of being compatibile out of the box with caching plugins? We should fix this before merging #10006.
- How can developers be educated and guided on implementing proper cache managements to prevent data leakage and ensure optimal performance?

Looking forward to read your feedback!

**Comment by luisherranz** [comment]
> The Interactivity API utilizes the data-* attribute to store user data, but it's important to note that cache plugins, such as WP-Optimize, cache the entire HTML code, including the data-* attributes.

It's probably worth noting that only `data-wp-context` (local state) can store user-specific information. The rest of the directives don't store state (just a reference to a callback) and therefore are not affected by this.

That said, you should never store user-specific information either in local context (`data-wp-context`) or in global context (`wp_store(...)`) on pages that are going to be cached. 

_This is not exactly a new rule, as it follows the same "you should never use user-specific information in the HTML of a page that is going to be cached" principle used by WordPress themes and plugins._

**Comment by Aljullu** [comment]
Thanks for writing this post, @gigitux!

> Should we take care of being compatibile out of the box with caching plugins?

Personally, I feel this is the most important question to answer because it will impact any new features we implement.

I don't have a strong opinion on it, but considering support for caching plugins has been a highly requested petition from the community, I would try to do it. However, we must keep in mind that it comes with a cost, in most cases, requiring more JS on the frontend and/or some extra requests to populate data. I definitely think it's worth for sites with caching enabled, but the drawback is that the costs are applied to all sites, even if they don't have caching enabled.

> Should we disable the cache per-block basis (not sure that it is possible)?

This could be a good interim solution to 'unblock' the PRs which are blocked by this decision, but I'm not sure up to which point it's scalable, if many blocks start disabling the cache, we might end up disabling caching in all pages.

> How can developers be educated and guided on implementing proper cache managements to prevent data leakage and ensure optimal performance?

You mean internal developers in WooCommerce or 3rd-party developers?

**Comment by gigitux** [comment]
> However, we must keep in mind that it comes with a cost, in most cases, requiring more JS on the frontend and/or some extra requests to populate data. I definitely think it's worth for sites with caching enabled, but the drawback is that the costs are applied to all sites, even if they don't have caching enabled.

The thing is, that is something that we want to avoid... To improve the performance, the best thing is to generate everything on the server side. Also, as you said, the drawback is that the costs are applied to all sites.

> This could be a good interim solution to 'unblock' the PRs which are blocked by this decision, but I'm not sure up to which point it's scalable, if many blocks start disabling the cache, we might end up disabling caching in all pages.

Agree with your thought.

> You mean internal developers in WooCommerce or 3rd-party developers?

I mean 3rd-party developers that develop cache plugins.

**Comment by luisherranz** [comment]
I don't have much say in this decision, but if bypassing cache with cart cookies is not a thing in WooCommerce sites, I think you should serve cachable HTML files for all the logged-out users.

It's true that it's a bit more challenging because you either have to remove those nice user-info customizations or figure out a nice UX to add them on the client after hydration. But big sites with lots of traffic will want to cache the product and product archive pages.

By the way, if you add the user info after hydration, it will work fine with the Products block client-side navigation. In other words, the user won't see any flickering when they paginate or change a filter. So you only would need to solve the UX of the first hydration.

**Reply by luisherranz** [reply]
Upon further consideration, if the [prevent_caching](https://github.com/woocommerce/woocommerce/blob/9c80e4d53fd7d24fd962468dd34219647f1a54ff/plugins/woocommerce/includes/class-wc-cache-helper.php/#L135-L148) feature is reliable, it may be possible to provide a cached version of the page to users who do not have the cart cookie, while serving a non-cached version to those who do. If the cache system respects these headers/constants, the version containing user information should never be cached.

The code would look something like this in each block that contains user info:

```php
if ( isset( WC()->cart ) ) {
	WC_Cache_Helper::prevent_caching();
}
```

> This could be a good interim solution to 'unblock' the PRs which are blocked by this decision, but I'm not sure up to which point it's scalable, if many blocks start disabling the cache, we might end up disabling caching in all pages.

The difference with this approach is that a cachable version of the page will still be generated whenever a new user (without an existing cart) visits that page. That cachable version won't contain any user info, it will be cached by the system (plugin, CDN) and it will be served to all subsequent users (even those with an existing cart).

The benefit of this approach is that sites without a cache system will serve properly server-side rendered HTML to users with existing carts. Sites with a cache system will not, but they can opt-in to server-side rendering if they want by adding the cart cookie to their _bypass cache_ settings.

The main drawback is that you still need to design the UX taking into account that sometimes the HTML will contain no user info and that needs to be populated after hydration.

**Reply by Aljullu** [reply]
I like this proposal. Also thanks for explaining it with so much detail, @luisherranz! :slightly_smiling_face: 

It feels like the best approach that gives us caching as in many circumstances as possible, but at the same time avoids showing invalid data to shoppers.

One follow-up question is whether we should update the Mini-Cart block to use this approach or not. I'm a bit on the verge between yes and no, but I would lean towards leaving it as-is. This way, pages which only show the Mini-Cart but no other blocks will still be able to render the cached version.

**Comment by gigitux** [comment]
@luisherranz thanks for your great suggestion! 

So, in summary, when a cache plugin is enabled, we will avoid caching the HTML that contains the user data. With the pseudo code, we can say that this will be the logic:

```php
if ( isset( WC()->cart ) ) {
	WC_Cache_Helper::prevent_caching();
    // Add user info...
} else {
    // Don't add user info...
}
```

Now, we should address this UX issue when the cache plugin is enabled: 


https://github.com/woocommerce/woocommerce-blocks/assets/4463174/0e6aa4f7-b5e0-41ec-9fd5-03ef8be9109f


Since the HTML that will be loaded does not contain user data, there will be a flickering effect caused by the delay in loading user data (handled by the JS side). How can we improve the UX for the final user? 

cc @shaunandrews  @jarekmorawski

**Reply by jarekmorawski** [reply]
Hi @gigitux! Do you mean the subtle jump when we load the button's state? How about a simple spinner?

https://github.com/woocommerce/woocommerce-blocks/assets/79307566/256ab2d2-7052-4428-b7f6-0d4f98a4553a

To remove the flickering, the button wouldn't need a fixed width (at least in the loading state), and the product card (image, name, price, and add-to-cart button) would require a fixed height, so the `View in cart` button doesn't push the next row of products down the screen.

**Reply by luisherranz** [reply]
The problem with the spinner is that it will appear on people with no cart:

```php
if ( isset( WC()->cart ) ) {
  // Not cached – Only users with a cart
  /** NO SPINNER **/
} else {
  // Cached – Users with a cart or no cart
  /** SPINNER **/
}
```

So you'd sacrifice a little bit the UX of the users with no cart, which I guess is going to be the majority.

Isn't there a way to smoothly transition from "Add to Cart" to "X in Cart" without displaying spinners so users without a cart don't see any change?

- Users with cart
  Initial `Add to cart` -> After hydration `X in the Cart`
- Users without cart
  Initial `Add to cart` -> After hydration `Add to cart`

**Reply by jarekmorawski** [reply]
It could help if we removed the ✔️ icon (not sure if it's helpful) and fixed the button's width. The jump wouldn't be as apparent anymore.

**Reply by luisherranz** [reply]
And is there something that could smooth the transition between the two texts to remove the flickering? Maybe a fade animation? Maybe even a spinner for a fixed time? (`Add to Cart` -> `SPINNER` (500ms) -> `X in the Cart`)

**Reply by jarekmorawski** [reply]
I probably misunderstood, but you said the spinner would show up even when the product isn't in the cart yet? In that case, how can we show the Add to Cart button first, then the spinner? When you look at the video I posted earlier, you'll see that the product that wasn't added yet has no spinner. Can we run the check when the user first enters the page and only show the spinner for products that haven't been cached yet?

**Reply by gigitux** [reply]
> but you said the spinner would show up even when the product isn't in the cart yet? 

Yes.

>  Can we run the check when the user first enters the page and only show the spinner for products that haven't been cached yet?

We can't do this kind of check because the entire page is cached.

**Reply by luisherranz** [reply]
Exactly.

The pattern is a bit tricky to understand, sorry 😅

Most of the users will visit the page without products added to the cart yet. If we want to prioritize their UX, the ideal option would be to show `Add to Cart` from the beginning:

_When the page is cached, WooCommerce only knows which products are added to the cart after fetching the Store API._

- `Add to cart` on the initial HTML
- Fetch the cart using the Store API
- There are no products in the cart, so still `Add to cart` (no change, therefore no flickering)

Some users will visit the cached page with products added to the cart. If we use `Add to Cart` on the initial HTML, that means users with products in the cart will see:

- `Add to cart` on the initial HTML
- Fetch the cart using the Store API
- Replace `Add to cart` with `X in the cart` (on the products that are added to the cart)

So I was wondering if there is a way to smooth the text transition between `Add to cart` and `X in the cart`.

Maybe with a fade…

- `Add to cart` on the initial HTML
- Fetch the cart using the Store API
- `Add to cart` fades out and `X in the cart` fades in (on the products that are added to the cart)

Maybe with a spinner…

- `Add to cart` on the initial HTML
- Fetch the cart using the Store API
- Replace `Add to cart` with a 500ms spinner to smooth the text transition (on the products that are added to the cart)
- Replace the spinner with `X in the cart`

Anyway, those are just my thoughts in case you choose to prioritize the UX of the users without products in the cart, but you should probably decide that first 🙂

**Reply by jarekmorawski** [reply]
Sounds good! I think it's fine to default to `Add to cart` being the default button state, then transition it to `X in cart` (I'd drop `the`) with some animation. I thought of a vertical slide because it's cleaner than a fade, but maybe it's too fancy? 🤔

https://github.com/woocommerce/woocommerce-blocks/assets/79307566/5ed7d5bf-c1f0-431f-b503-1a9ecdd60ded

**Reply by luisherranz** [reply]
Ohhh, that's so nice!! I love it.

**Reply by gigitux** [reply]
Looks great! Thank you! 🚀

**Reply by gigitux** [reply]
I'm working on this animation. So, is it okay that the view cart link will not be visible anymore, right? @jarekmorawski  
![image](https://github.com/woocommerce/woocommerce-blocks/assets/4463174/4d49719e-aaa6-46c6-b046-14365696ea4b)

**Reply by gigitux** [reply]
@jarekmorawski This is the PR that introduce the new animation: https://github.com/woocommerce/woocommerce-blocks/pull/10351

**Reply by jarekmorawski** [reply]
Let's keep the view cart link (I forgot to include it in the mockup). Can we place it beside the add to cart button so it doesn't increase the height of the product card so much?

**Reply by gigitux** [reply]
I updated #10351! Let me know what you think about it!

**Comment by ResponsiveFlow** [comment]
I have two questions:
1. If you plan to serve an uncached version of the page to those users who have a cart cookie, does it mean that the server resources required will substantially increase?
2. Do you plan to extend the Add to cart with interactivity API to the single product template?

**Reply by gigitux** [reply]
> If you plan to serve an uncached version of the page to those users who have a cart cookie, does it mean that the server resources required will substantially increase?

If a cache plugin is configurated, we will serve the cached page, and the buttons will be hydrated and updated via JS. We are figuring out the best UX. 

> Do you plan to extend the Add to cart with interactivity API to the single product template?

At least, for now, no. But if you have reasons why we should do this, drop a comment in this [GH discussion](https://github.com/woocommerce/woocommerce-blocks/discussions/10140).

**Reply by ResponsiveFlow** [reply]
I've added the comment to the GitHub discussion woocommerce/woocommerce-blocks#10140

**Reply by gigitux** [reply]
Thanks!

**Comment by gigitux** [comment]
I'm closing this issue given that we have been able to figure out a solution
